tests/cpp/tests/test_nodes_mathematical.cpp

```diff
@@ -567,59 +456,346 @@ TEMPLATE_TEST_CASE("NaryOpNode", "", functional::max<double>, functional::min<do
+                list_ptr->exchange(state, 2, 4);
+                list_ptr->propagate(state);
+                max_ptr->propagate(state);
+                min_ptr->propagate(state);
+
+                THEN("The min and max are as expected") {
+                    CHECK(max_ptr->view(state)[0] == 4);
+                    CHECK(min_ptr->view(state)[0] == 1);
                 }
             }
 
-            AND_WHEN("We modify the 2 lists with single and overlapping changes") {
-                a_ptr->exchange(state, 0, 1);
-                b_ptr->exchange(state, 1, 2);
-
-                a_ptr->propagate(state);
-                b_ptr->propagate(state);
-                p_ptr->propagate(state);
+            WHEN("We swap out the max value for something smaller [ 1 2 0 | 3 4 ]") {
+                list_ptr->exchange(state, 2, 3);
+                list_ptr->propagate(state);
+                max_ptr->propagate(state);
+                min_ptr->propagate(state);
 
-                THEN("The product has the values we expect") {
-                    for (int i = 0; i < p_ptr->size(state); ++i) {
-                        CHECK(p_ptr->view(state)[i] ==
-                              func(a_ptr->view(state)[i],
-                                   func(b_ptr->view(state)[i], c_ptr->view(state)[i])));
-                    }
+                THEN("The min and max are as expected") {
+                    CHECK(max_ptr->view(state)[0] == 2);
+                    CHECK(min_ptr->view(state)[0] == 0);
                 }
             }
         }
     }
 
-    GIVEN("A vector<Node*> of constants") {
-        auto nodes = std::vector<ArrayNode*>{
-                graph.emplace_node<ConstantNode>(5), graph.emplace_node<ConstantNode>(6),
-                graph.emplace_node<ConstantNode>(7), graph.emplace_node<ConstantNode>(8)};
-
-        WHEN("We construct a NaryOpNode from it") {
-            auto p_ptr = graph.emplace_node<NaryOpNode<TestType>>(nodes);
-
-            THEN("The shape is also a scalar") {
-                CHECK(p_ptr->ndim() == 0);
-                CHECK(p_ptr->size() == 1);
-            }
-
-            WHEN("We make a state") {
-                auto state = graph.initialize_state();
+    GIVEN("A dynamic node") {
+        auto a_ptr = graph.emplace_node<SetNode>(4);
 
-                THEN("The product has the value and shape we expect") {
-                    CHECK(p_ptr->size(state) == 1);
-                    CHECK(p_ptr->shape(state).size() == 0);
-                    CHECK(p_ptr->view(state)[0] == func(5, func(6, func(7, 8))));
-                }
-            }
+        THEN("We cannot construct a reduce node without an initial value") {
+            CHECK_THROWS(graph.emplace_node<MaxNode>(a_ptr));
+            CHECK_THROWS(graph.emplace_node<MinNode>(a_ptr));
+            CHECK(a_ptr->successors().size() == 0);  // no side effects
         }
     }
 }
 
-TEST_CASE("ProdNode") {
+TEST_CASE("ReduceNode - ProdNode") {
```

Comment:
Could add a `[ReduceNode]` tag to these

Reply:
I wanted them visible in the output when there's an error and tags for whatever reason don't show up there. Agree we could add tags, but IMO that's a followup PR for all of the tests - currently they are all untagged.

Reply:
Sure! Just noting it would be nice to be able to run all the `ReduceNode` tests together still. Wasn't recommending removing the `ReduceNode` from the main title.